pkg/util/converter/prom.go

```diff
@@ -925,6 +946,182 @@ func readStream(iter *jsonitere.Iterator) backend.DataResponse {
 	return rsp
 }
 
+func readCategorizedStream(iter *jsonitere.Iterator) backend.DataResponse {
+	rsp := backend.DataResponse{}
+
+	labelsField := data.NewFieldFromFieldType(data.FieldTypeJSON, 0)
+	labelsField.Name = "__labels" // avoid automatically spreading this by labels
+
+	labelTypesField := data.NewFieldFromFieldType(data.FieldTypeJSON, 0)
+	labelTypesField.Name = "__labelTypes" // avoid automatically spreading this by labels
+
+	timeField := data.NewFieldFromFieldType(data.FieldTypeTime, 0)
+	timeField.Name = "Time"
+
+	lineField := data.NewFieldFromFieldType(data.FieldTypeString, 0)
+	lineField.Name = "Line"
+
+	// Nanoseconds time field
+	tsField := data.NewFieldFromFieldType(data.FieldTypeString, 0)
+	tsField.Name = "TS"
+
+	labels := data.Labels{}
+	labelMap, err := labelsToMap(labels)
+	if err != nil {
+		return backend.DataResponse{Error: err}
+	}
+
+	for more, err := iter.ReadArray(); more; more, err = iter.ReadArray() {
+		if err != nil {
+			rspErr(err)
+		}
+
+	l1Fields:
+		for l1Field, err := iter.ReadObject(); ; l1Field, err = iter.ReadObject() {
+			if err != nil {
+				return rspErr(err)
+			}
+			switch l1Field {
+			case "stream":
+				// we need to clear `labels`, because `iter.ReadVal`
+				// only appends to it
+				labels := data.Labels{}
+				if err = iter.ReadVal(&labels); err != nil {
+					return rspErr(err)
+				}
+
+				if labelMap, err = labelsToMap(labels); err != nil {
+					return rspErr(err)
+				}
+
+			case "values":
+				for more, err := iter.ReadArray(); more; more, err = iter.ReadArray() {
+					if err != nil {
```

Comment:
I would love to see this extracted to a function so this function is easier to read and less spaghetty.